lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py

```diff
@@ -112,55 +113,93 @@ def __init__(self):
         self.ue_ipv6_block = "fdee:0005:006c::/64"
         self.agw_ipv6 = "3001::10"
 
-    def exec_command(self, command):
+    def exec_command(self, command, capture_output=False):
         """
         Run a command remotely on magma_trfserver VM.
 
         Args:
             command: command (str) to be executed on remote host
                 e.g. 'sed -i \'s/str1/str2/g\' /usr/local/bin/traffic_server.py'
+            capture_output: Sets the `capture_output` flag in `subprocess.run`
 
         Returns:
-            Shell command execution output
+            Shell command return code and stdout
         """
-        data = self._cmd_data
-        data["command"] = '"' + command + '"'
-        param_list = shlex.split(self._command.format(**data))
-        return subprocess.call(
+        self._cmd_data["command"] = f'"{command}"'
+        param_list = shlex.split(self._command.format(**self._cmd_data))
+        return subprocess.run(
             param_list,
             shell=False,
-            stdout=subprocess.DEVNULL,
-            stderr=subprocess.DEVNULL,
+            capture_output=capture_output,
         )
 
+    def dump_leases(self):
+        """Dump DHCP leases in TRF server VM"""
+        return self.exec_command("dumpleases", capture_output=True).stdout
+
+    def clear_leases(self):
+        """Remove all DHCP leases in TRF server VM"""
+        cmd = 'systemctl stop udhcpd.service && ' \
+              'rm -f /var/lib/misc/udhcpd.leases && ' \
+              'systemctl start udhcpd.service'
+        return self.exec_command(f"sudo bash -c '{cmd}'").returncode
+
+    @property
+    def n_leases(self):
+        """Count the total number of leases in TRF server VM"""
+        return len(self.dump_leases().decode("utf-8").split("\n")) - 2
+
+    @property
+    def n_expired_leases(self):
+        """Count number of expired leases in TRF server VM"""
+        return self.dump_leases().decode("utf-8").count("expired")
+
+    def check_attached_leases(self, expected_leases, timeout=5):
+        """Wait for up to timeout seconds for expected number of leases"""
+        i = 0
+        while expected_leases != self.n_leases:
+            if i == timeout - 1:
+                assert False, "IP not assigned to UE"
+            time.sleep(1)
+            i += 1
+
+    def check_detached_leases(self, expected_leases, timeout=5):
+        """Wait for up to timeout seconds for expected number of leases"""
+        i = 0
+        while expected_leases != self.n_expired_leases:
```

Comment:
In my tests the loop was never entered - did you had cases where it took some time until all leases were expired? I think the loop/wait is still reasonable - just wondering.

Reply:
I believe this happened once for me. It would be a relatively rare case, but the leases are first removed in redis and then on the traffic server. It would not be out of sync for a long time, but in the test code the release loop might could exactly end when the last redis entry is removed but the traffic server is still out of sync. You might be able to trigger that case more often by removing the 5 second sleep from the release loop in the test code.